File: generate_session.py
"""
Génère une session Telethon et l'envoie dans votre chat Telegram privé.

Usage (dans le Shell Replit) :
    python generate_session.py

Étapes :
  1. Entrez votre numéro de téléphone  (+225XXXXXXXXXX)
  2. Telegram vous envoie un code de vérification
  3. Tapez le code avec le préfixe "aa"  →  ex : aa12345
  4. La session est affichée dans le Shell ET envoyée dans votre chat Telegram
"""

import asyncio
import os
import sys

import httpx
from telethon import TelegramClient
from telethon.sessions import StringSession

# ── Récupérer les variables depuis config.py ──────────────────
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from config import TELETHON_API_ID, TELETHON_API_HASH, BOT_TOKEN, ADMINS


async def send_to_telegram(session_str: str):
    """Envoie la session dans le chat privé du premier admin."""
    if not ADMINS or not BOT_TOKEN:
        return
    admin_id = ADMINS[0]
    text = (
        "🔑 *Session Telethon générée avec succès*\n\n"
        "Copiez cette chaîne et utilisez-la comme variable d'environnement "
        "`TELETHON_SESSION` sur Render.com ou dans votre `.env` :\n\n"
        f"`{session_str}`\n\n"
        "⚠️ *Ne partagez jamais cette chaîne avec personne.*"
    )
    try:
        async with httpx.AsyncClient(timeout=15) as http:
            r = await http.post(
                f"https://api.telegram.org/bot{BOT_TOKEN}/sendMessage",
                json={"chat_id": admin_id, "text": text, "parse_mode": "Markdown"},
            )
            if r.status_code == 200:
                print(f"✅ Session envoyée dans votre chat Telegram (ID: {admin_id})")
            else:
                print(f"⚠️  Envoi Telegram échoué ({r.status_code}): {r.text}")
    except Exception as e:
        print(f"⚠️  Erreur envoi Telegram: {e}")


async def main():
    print()
    print("═" * 62)
    print("   GÉNÉRATEUR DE SESSION TELETHON")
    print("═" * 62)
    print()

    phone = input("📱 Numéro de téléphone (avec indicatif, ex: +22500000000): ").strip()
    if not phone.startswith("+"):
        phone = "+" + phone

    client = TelegramClient(StringSession(), TELETHON_API_ID, TELETHON_API_HASH)
    await client.connect()

    print(f"\n⏳ Envoi du code de vérification à {phone} …")
    try:
        await client.send_code_request(phone)
    except Exception as e:
        print(f"❌ Impossible d'envoyer le code : {e}")
        await client.disconnect()
        return

    print("\n📨 Un code vient d'arriver dans votre application Telegram.")
    print("   Entrez-le ici avec le préfixe  'aa'  (exemple : aa12345)")
    raw = input("🔑 Code : ").strip()

    if not raw.lower().startswith("aa"):
        print("❌ Le code doit commencer par 'aa'  (exemple : aa12345). Abandonné.")
        await client.disconnect()
        return

    code = raw[2:].strip()  # supprimer le préfixe "aa"

    try:
        await client.sign_in(phone, code)
    except Exception as e:
        err = str(e).lower()
        if "two-step" in err or "password" in err or "2fa" in err:
            pwd = input("🔐 2FA activé — entrez votre mot de passe Telegram : ").strip()
            try:
                await client.sign_in(password=pwd)
            except Exception as e2:
                print(f"❌ Erreur 2FA : {e2}")
                await client.disconnect()
                return
        else:
            print(f"❌ Erreur de connexion : {e}")
            await client.disconnect()
            return

    session_str = client.session.save()
    await client.disconnect()

    print()
    print("═" * 62)
    print("✅  SESSION GÉNÉRÉE AVEC SUCCÈS")
    print("═" * 62)
    print()
    print(session_str)
    print()
    print("═" * 62)
    print("Copiez la chaîne ci-dessus dans la variable TELETHON_SESSION")
    print("sur Render.com (ou dans vos secrets Replit).")
    print("═" * 62)
    print()

    # Envoyer également dans le chat Telegram
    await send_to_telegram(session_str)
    print()


asyncio.run(main())

File: generate-session.js
// ============================================================
// À EXÉCUTER UNE SEULE FOIS, EN LOCAL, SUR TON PROPRE ORDINATEUR.
// NE JAMAIS EXÉCUTER CE SCRIPT SUR RENDER NI LE PARTAGER À QUELQU'UN D'AUTRE.
//
// Ce script connecte ton compte Telegram (numéro de téléphone) via l'API
// officielle MTProto et génère une "session string". Le code de vérification
// Telegram te sera envoyé directement dans ton appli Telegram (ou par SMS) :
// SEUL TOI dois le saisir ici, jamais à personne d'autre, jamais à Claude.
//
// Utilisation :
//   1. npm install telegram
//   2. node generate-session.js
//   3. Réponds aux questions (numéro, code reçu, mot de passe 2FA si tu en as un)
//   4. Copie la valeur affichée à la fin dans la variable d'environnement
//      TG_SESSION sur Render (Dashboard → ton service → Environment)
// ============================================================

const { TelegramClient } = require("telegram");
const { StringSession }  = require("telegram/sessions");
const readline = require("readline");

// Ces deux valeurs viennent de https://my.telegram.org (mêmes valeurs que
// TG_API_ID / TG_API_HASH que tu configureras sur Render).
const API_ID   = parseInt(process.env.TG_API_ID   || "29177661", 10);
const API_HASH = process.env.TG_API_HASH || "a8639172fa8d35dbfd8ea46286d349ab";

function ask(question) {
  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
  return new Promise(resolve => rl.question(question, answer => { rl.close(); resolve(answer.trim()); }));
}

(async () => {
  console.log("=== Connexion à ton compte Telegram (une seule fois) ===\n");
  const client = new TelegramClient(new StringSession(""), API_ID, API_HASH, {
    connectionRetries: 5,
  });

  await client.start({
    phoneNumber: async () => await ask("📱 Ton numéro Telegram, avec indicatif (ex: +22995501564) : "),
    password:    async () => await ask("🔒 Mot de passe 2FA (laisse vide si tu n'en as pas configuré) : "),
    phoneCode:   async () => await ask("💬 Code reçu dans Telegram (ou par SMS) : "),
    onError:     (err) => console.error(err),
  });

  const sessionString = client.session.save();

  console.log("\n✅ Connecté avec succès !\n");
  console.log("Copie EXACTEMENT cette valeur dans la variable d'environnement");
  console.log("TG_SESSION sur Render (Dashboard → ton service → Environment) :\n");
  console.log(sessionString);
  console.log("\n⚠️  Ne partage cette valeur avec PERSONNE (pas même moi) : elle donne un");
  console.log("    accès complet à ton compte Telegram, comme un mot de passe.\n");

  await client.disconnect();
  process.exit(0);
})();
